onnxruntime/core/providers/shared_library/provider_bridge_provider.cc

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This is the provider DLL side of the provider API to let providers be built as a DLL
+
+#include "provider_author.h"
+#include <assert.h>
+#include <mutex>
+#include <iostream>  // For std::cout used in a stub
+
+onnxruntime::ProviderHost* g_host{};
+
+#define PROVIDER_NOT_IMPLEMENTED ORT_THROW("Unimplemented shared library provider method");
+
+namespace onnxruntime {
+
+void SetProviderHost(ProviderHost& host) {
+  g_host = &host;
+}
+
+struct OnUnloadFunction {
+  OnUnloadFunction(std::function<void()> function) : function_(std::move(function)) {}
+
+  std::function<void()> function_;
+  bool enabled_{true};
+};
+
+static std::unique_ptr<std::vector<std::unique_ptr<OnUnloadFunction>>> s_run_on_unload_;
+
+RunOnUnload::RunOnUnload(std::function<void()> deleter) {
+  static std::mutex mutex;
+  std::lock_guard<std::mutex> guard{mutex};
+  if (!s_run_on_unload_)
+    s_run_on_unload_ = onnxruntime::make_unique<std::vector<std::unique_ptr<OnUnloadFunction>>>();
+  auto unload_function = std::make_unique<OnUnloadFunction>(std::move(deleter));
+  enabled_ = &unload_function->enabled_;
+  s_run_on_unload_->push_back(std::move(unload_function));
+}
+
+RunOnUnload::~RunOnUnload() {
+  *enabled_ = false;  // If the thread_local gets destroyed, then disalble the delete function
+}
```

Comment:
nit: 'disalble'

Reply:
Yeah, this could go either way, but the if statement that uses it is 'if(enabled)' vs if(!disabled). So that lose a point there.

Reply:
Hah, I first thought you wanted me to rename it to 'disabled' but then saw my typo. Fixed